File: GameLibrary/apps/cc/applets/typing/LearnToTypeBuilder.java
package cc.applets.typing;

import java.awt.Graphics;
import java.awt.Point;
import java.awt.event.FocusEvent;
import java.awt.event.FocusListener;
import java.awt.event.KeyEvent;
import java.awt.event.KeyListener;
import java.awt.event.MouseEvent;
import java.awt.event.MouseListener;
import java.awt.event.MouseMotionListener;
import java.awt.image.BufferedImage;
import java.io.File;
import java.io.FileNotFoundException;
import java.util.HashMap;
import java.util.Map;

import javax.swing.JApplet;

import cc.lib.game.AGraphics;
import cc.lib.game.AImage;
import cc.lib.game.GColor;
import cc.lib.game.GRectangle;
import cc.lib.game.Justify;
import cc.lib.math.Vector2D;
import cc.lib.swing.AWTFrame;
import cc.lib.swing.AWTGraphics;
import cc.lib.utils.FileUtils;
import cc.lib.utils.Reflector;

public class LearnToTypeBuilder extends JApplet
        implements MouseListener,
        MouseMotionListener,
        KeyListener,
        FocusListener {

    private Graphics G = null;
    private AGraphics AG = null;

    public static void main(String [] args) {
        final LearnToTypeBuilder app = new LearnToTypeBuilder();
        AWTFrame frame = new AWTFrame("Learn to Type") {
            @Override
            protected void onWindowClosing() {
                try {
                    File resources = new File("GameLibrary/resources");
                    if (!resources.isDirectory())
                        throw new Exception("resources folder:\n '" + resources + "'\n not found in working directory:\n " + new File(".").getCanonicalPath());
                    Reflector.serializeToFile(app.keyboard, new File(resources, keyBoardFile.getName()));
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }
        };
        app.setCursor( app.getToolkit().createCustomCursor(
                new BufferedImage( 1, 1, BufferedImage.TYPE_INT_ARGB ),
                new Point(),
                null ) );
        frame.add(app);
        File settings = FileUtils.getOrCreateSettingsDirectory(LearnToTypeBuilder.class);
        frame.setPropertiesFile(new File(settings, "app.properties"));
        if (!frame.restoreFromProperties())
            frame.centerToScreen(800, 600);
        app.init();
        app.start();
    }

    Map<Integer, GRectangle> keyboard = new HashMap<>();
    static File keyBoardFile = new File("keyboard.txt");

    @Override
    public void init() {
        addFocusListener(this);
        addKeyListener(this);
        addMouseListener(this);
        addMouseMotionListener(this);
        try {
            Reflector.deserializeFromFile(keyBoardFile);
        } catch (FileNotFoundException e) {
            // ignore
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @Override
    public final void paint(Graphics g) {

        if (g != G) {
            G = g;
            AG = new AWTGraphics(g, this);
            setup(AG);
        }

        draw(AG);
    }

    int keybdId = -1;

    protected void setup(AGraphics g) {
        if (keybdId < 0)
            keybdId = g.loadImage("small_keyboard.png");
    }

    KeyEvent pressed = null;
    GRectangle rect = null;
    boolean dragging = false;
    int mouseX=0, mouseY=0;

    protected void draw(AGraphics g) {

        g.clearScreen(GColor.GREEN);
        if (keybdId >= 0) {
            AImage img = g.getImage(keybdId);
            g.drawImage(keybdId, 0, 0, g.getViewportWidth(), (float)g.getViewportWidth()/img.getAspect());
        }

        for (Map.Entry<Integer, GRectangle> e : keyboard.entrySet()) {
            g.setColor(GColor.WHITE);
            g.drawFilledRoundedRect(e.getValue(), 10);
            g.setColor(GColor.BLACK);
            String key = KeyEvent.getKeyText(e.getKey());
            g.drawJustifiedString(e.getValue().getCenter(), Justify.CENTER, Justify.CENTER, key);
        }

        g.setColor(GColor.YELLOW);
        g.setLineWidth(3);
        if (dragging) {
            g.drawRect(rect);
        }

        if (pressed != null) {
            g.setTextHeight(32);
            g.setColor(GColor.BLUE);
            g.setTextStyles(AGraphics.TextStyle.BOLD);
            g.drawJustifiedString(g.getViewportWidth()/2, g.getViewportHeight()*3/4, KeyEvent.getKeyText(pressed.getKeyCode()));
        }

        // draw crosshais
        g.setColor(GColor.LIGHT_GRAY);
        final int SIZE = 10;
        g.drawLine(mouseX-SIZE, mouseY, mouseX+SIZE, mouseY);
        g.drawLine(mouseX, mouseY-SIZE, mouseX, mouseY+SIZE);
    }

    @Override
    public void keyTyped(KeyEvent e) {
        pressed = e;
        repaint();
    }

    @Override
    public void keyPressed(KeyEvent e) {
        pressed = e;
        repaint();
    }

    @Override
    public void keyReleased(KeyEvent e) {
        pressed = e;
        repaint();
    }

    @Override
    public void mouseClicked(MouseEvent e) {

    }



    @Override
    public void mousePressed(MouseEvent e) {
        if (pressed != null) {
            Vector2D v = new Vector2D(mouseX, mouseY);
            rect = new GRectangle(v, v);
            repaint();
        }
    }

    @Override
    public void mouseReleased(MouseEvent e) {
        if (dragging) {
            keyboard.put(pressed.getKeyCode(), rect);
            repaint();
        }
    }

    @Override
    public void mouseEntered(MouseEvent e) {

    }

    @Override
    public void mouseExited(MouseEvent e) {

    }

    @Override
    public void mouseDragged(MouseEvent e) {
        mouseMoved(e);
        if (pressed != null) {
            dragging = true;
            Vector2D v = new Vector2D(mouseX, mouseY);
            rect.setEnd(v);
        }
    }

    final int GRID = 5;

    @Override
    public void mouseMoved(MouseEvent e) {
        mouseX = (e.getX()/GRID) * GRID;
        mouseY = (e.getY()/GRID) * GRID;
        repaint();
    }

    @Override
    public void focusGained(FocusEvent e) {

    }

    @Override
    public void focusLost(FocusEvent e) {

    }
}

File: SwingApps/apps/cc/applets/typing/LearnToTypeBuilder.kt
package cc.applets.typing

import cc.lib.game.AGraphics
import cc.lib.game.GColor
import cc.lib.game.GRectangle
import cc.lib.game.Justify
import cc.lib.math.Vector2D
import cc.lib.swing.AWTFrame
import cc.lib.swing.AWTGraphics
import cc.lib.utils.FileUtils
import cc.lib.utils.Reflector
import java.awt.Graphics
import java.awt.Point
import java.awt.event.*
import java.awt.image.BufferedImage
import java.io.File
import java.io.FileNotFoundException
import javax.swing.JApplet

class LearnToTypeBuilder : JApplet(), MouseListener, MouseMotionListener, KeyListener, FocusListener {
	private var G: Graphics? = null
	private var AG: AGraphics? = null
	var keyboard: MutableMap<Int, GRectangle?> = HashMap()
	override fun init() {
		addFocusListener(this)
		addKeyListener(this)
		addMouseListener(this)
		addMouseMotionListener(this)
		try {
			Reflector.deserializeFromFile<Any>(keyBoardFile)
		} catch (e: FileNotFoundException) {
			// ignore
		} catch (e: Exception) {
			e.printStackTrace()
		}
	}

	override fun paint(g: Graphics) {
		if (g !== G) {
			G = g
			with(AWTGraphics(g, this)) {
				AG = this
				setup(this)
			}
		}
		draw(AG)
	}

	var keybdId = -1
	protected fun setup(g: AGraphics) {
		if (keybdId < 0) keybdId = g.loadImage("small_keyboard.png")
	}

	var pressed: KeyEvent? = null
	var rect: GRectangle? = null
	var dragging = false
	var mouseX = 0
	var mouseY = 0
	protected fun draw(g: AGraphics?) {
		g!!.clearScreen(GColor.GREEN)
		if (keybdId >= 0) {
			val img = g.getImage(keybdId)
			g.drawImage(keybdId, 0f, 0f, g.viewportWidth.toFloat(), g.viewportWidth.toFloat() / img.aspect)
		}
		for ((key1, value) in keyboard) {
			g.color = GColor.WHITE
			g.drawFilledRoundedRect(value, 10f)
			g.color = GColor.BLACK
			val key = KeyEvent.getKeyText(key1)
			g.drawJustifiedString(value!!.center, Justify.CENTER, Justify.CENTER, key)
		}
		g.color = GColor.YELLOW
		g.setLineWidth(3f)
		if (dragging) {
			g.drawRect(rect)
		}
		if (pressed != null) {
			g.textHeight = 32f
			g.color = GColor.BLUE
			g.setTextStyles(AGraphics.TextStyle.BOLD)
			g.drawJustifiedString((g.viewportWidth / 2).toFloat(), (g.viewportHeight * 3 / 4).toFloat(), KeyEvent.getKeyText(pressed!!.keyCode))
		}

		// draw crosshais
		g.color = GColor.LIGHT_GRAY
		val SIZE = 10
		g.drawLine((mouseX - SIZE).toFloat(), mouseY.toFloat(), (mouseX + SIZE).toFloat(), mouseY.toFloat())
		g.drawLine(mouseX.toFloat(), (mouseY - SIZE).toFloat(), mouseX.toFloat(), (mouseY + SIZE).toFloat())
	}

	override fun keyTyped(e: KeyEvent) {
		pressed = e
		repaint()
	}

	override fun keyPressed(e: KeyEvent) {
		pressed = e
		repaint()
	}

	override fun keyReleased(e: KeyEvent) {
		pressed = e
		repaint()
	}

	override fun mouseClicked(e: MouseEvent) {}
	override fun mousePressed(e: MouseEvent) {
		if (pressed != null) {
			val v = Vector2D(mouseX.toFloat(), mouseY.toFloat())
			rect = GRectangle(v, v)
			repaint()
		}
	}

	override fun mouseReleased(e: MouseEvent) {
		if (dragging) {
			keyboard[pressed!!.keyCode] = rect
			repaint()
		}
	}

	override fun mouseEntered(e: MouseEvent) {}
	override fun mouseExited(e: MouseEvent) {}
	override fun mouseDragged(e: MouseEvent) {
		mouseMoved(e)
		if (pressed != null) {
			dragging = true
			val v = Vector2D(mouseX.toFloat(), mouseY.toFloat())
			rect!!.setEnd(v)
		}
	}

	val GRID = 5
	override fun mouseMoved(e: MouseEvent) {
		mouseX = e.x / GRID * GRID
		mouseY = e.y / GRID * GRID
		repaint()
	}

	override fun focusGained(e: FocusEvent) {}
	override fun focusLost(e: FocusEvent) {}

	companion object {
		@JvmStatic
		fun main(args: Array<String>) {
			val app = LearnToTypeBuilder()
			val frame: AWTFrame = object : AWTFrame("Learn to Type") {
				override fun onWindowClosing() {
					try {
						val resources = File("GameLibrary/resources")
						if (!resources.isDirectory) throw Exception("""resources folder:
 '$resources'
 not found in working directory:
 ${File(".").canonicalPath}""")
						Reflector.serializeToFile<Any>(app.keyboard, File(resources, keyBoardFile.name))
					} catch (e: Exception) {
						e.printStackTrace()
					}
				}
			}
			app.cursor = app.toolkit.createCustomCursor(
				BufferedImage(1, 1, BufferedImage.TYPE_INT_ARGB),
				Point(),
				null)
			frame.add(app)
			val settings = FileUtils.getOrCreateSettingsDirectory(LearnToTypeBuilder::class.java)
			frame.setPropertiesFile(File(settings, "app.properties"))
			if (!frame.restoreFromProperties()) frame.centerToScreen(800, 600)
			app.init()
			app.start()
		}

		var keyBoardFile = File("keyboard.txt")
	}
}